Spark docs (#5796)
huggingface/datasets: docs/source/use_with_spark.mdx

# Use with Spark

> This document is a quick introduction to using `datasets` with Spark, with a particular focus on how to get
> a data loader to train models from a Spark DataFrame.

Reviewer: ```suggestion
This document is a quick introduction to using 🤗 Datasets with Spark, with a particular focus on how to load a Spark DataFrame into a [`Dataset`] object.
```

From there, you have fast access to any element and you can use it as a data loader to train models.

## Load from Spark

A [`Dataset`] object is a wrapper of an Arrow table, which allows fast reads from arrays in the dataset to PyTorch, TensorFlow and JAX tensors.

> The Arrow table is memory mapped from disk, which allows to load datasets bigger than RAM.

Reviewer: ```suggestion
The Arrow table is memory mapped from disk, which can load datasets bigger than your available RAM.
```

You can get a [`Dataset`] from a Spark DataFrame using [`Dataset.from_spark`]:

```py
>>> from datasets import Dataset
>>> df = spark.createDataFrame(
...     data=[[1, "Elia"], [2, "Teo"], [3, "Fang"]],
...     columns=["id", "name"],
... )
>>> ds = Dataset.from_spark(df)
```

The Spark workers write the dataset on disk in a cache directory as Arrow files, and the [`Dataset`] is loaded from there.

### Caching

> The resulting [`Dataset`] is cached: if you call [`Dataset.from_spark`] multiple times on the say DataFrame it won't re-run the Spark

Reviewer 2: type: say -> same

Reviewer 2: *typo :) also below

> The resulting [`Dataset`] is cached: if you call [`Dataset.from_spark`] multiple times on the same DataFrame it won't re-run the Spark

Reviewer: ```suggestion
The resulting [`Dataset`] is cached; if you call [`Dataset.from_spark`] multiple times on the same DataFrame it won't re-run the Spark
```

> job that writes the dataset as Arrow fiels on disk.

Reviewer 2: type: fiels -> fields

> You can set the location by passing `cache_dir=` to [`Dataset.from_spark`].

Reviewer: ```suggestion
You can set the cache location by passing `cache_dir=` to [`Dataset.from_spark`].
```

Make sure to use a disk that is available to both your workers and your current machine (the driver).

> Note that in a different session, the Spark DataFrame doesn't have the same [semantic hash](https://spark.apache.org/docs/3.2.0/api/python/reference/api/pyspark.sql.DataFrame.semanticHash.html) and therefore it would use a new cache.

Reviewer: ```suggestion
<Tip warning={true}>

In a different session, a Spark DataFrame doesn't have the same [semantic hash](https://spark.apache.org/docs/3.2.0/api/python/reference/api/pyspark.sql.DataFrame.semanticHash.html), and it will rerun a Spark job and store it in a new cache.

</Tip>
```
